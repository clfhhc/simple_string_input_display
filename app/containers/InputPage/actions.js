import {
  INPUT_PAGE_CHANGE_INPUT,
  INPUT_PAGE_INSERT,
  INPUT_PAGE_FETCH_HISTORY,
  INPUT_PAGE_FETCHED,
  INPUT_PAGE_INSERTED,
  INPUT_PAGE_ERROR,
  INPUT_PAGE_LOADING,
} from './constants';

export function changeInput(stringToInput) {
  return {
    type: INPUT_PAGE_CHANGE_INPUT,
    stringToInput,
  };
}

export function insertInput() {
  return {
    type: INPUT_PAGE_INSERT,
  };
}

export function fetchHistory() {
  return {
    type: INPUT_PAGE_FETCH_HISTORY,
  };
}

export function updateHistory(inputHistory) {
  return {
    type: INPUT_PAGE_FETCHED,
    inputHistory,
  };
}

export function inputInserted() {
  return {
    type: INPUT_PAGE_INSERTED,
  };
}

export function fetchedError(error) {
  return {
    type: INPUT_PAGE_ERROR,
    error,
  };
}

export function fetchLoading() {
  return {
    type: INPUT_PAGE_LOADING,
  };
}
