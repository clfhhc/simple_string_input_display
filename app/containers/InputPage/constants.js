export const INPUT_PAGE_CHANGE_INPUT = 'app/InputPage/INPUT_PAGE_CHANGE_INPUT';
export const INPUT_PAGE_INSERT = 'app/InputPage/INPUT_PAGE_INSERT';
export const INPUT_PAGE_FETCH_HISTORY =
  'app/InputPage/INPUT_PAGE_FETCH_HISTORY';
export const INPUT_PAGE_LOADING = 'app/InputPage/INPUT_PAGE_LOADING';
export const INPUT_PAGE_ERROR = 'app/InputPage/INPUT_PAGE_ERROR';
export const INPUT_PAGE_FETCHED = 'app/InputPage/INPUT_PAGE_FETCHED';
