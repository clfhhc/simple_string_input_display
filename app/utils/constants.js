export const RESTART_ON_REMOUNT = '@@saga-injector/restart-on-remount';
export const DAEMON = '@@saga-injector/daemon';
export const ONCE_TILL_UNMOUNT = '@@saga-injector/once-till-unmount';
export const serverUrl = 'http://localHost:3000';
