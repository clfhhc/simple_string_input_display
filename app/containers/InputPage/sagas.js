import {
  call,
  put,
  cancel,
  fork,
  select,
  takeLatest,
} from 'redux-saga/effects';
import { serverUrl } from 'utils/constants';
import { getFromUrl, postToUrl } from './apis';
import { INPUT_PAGE_INSERT, INPUT_PAGE_FETCH_HISTORY } from './constants';
import { updateHistory, fetchLoading, fetchedError } from './actions';
import { makeSelectStringToInput } from './selectors';

const inputUrl = `${serverUrl}/input`;

export function* fetchIsLoading() {
  yield put(fetchLoading());
}

export function* requestHistoryFromServer() {
  try {
    const task = yield fork(fetchIsLoading);
    const { inputHistory } = yield call(getFromUrl, inputUrl);
    yield cancel(task);
    yield put(updateHistory(inputHistory));
  } catch (err) {
    yield put(fetchedError(err));
  }
}

export function* insertStringToServer() {
  const stringToInput = yield select(makeSelectStringToInput());
  try {
    const task = yield fork(fetchIsLoading);
    yield call(postToUrl, inputUrl, { stringToInput });
    yield cancel(task);
    // yield call(requestHistoryFromServer);
  } catch (err) {
    yield put(fetchedError(err));
  }
}

function* insertInput() {
  yield takeLatest(INPUT_PAGE_INSERT, insertStringToServer);
}

function* fetchHistory() {
  yield takeLatest(INPUT_PAGE_FETCH_HISTORY, requestHistoryFromServer);
}

export default {
  insertInput,
  fetchHistory,
};
