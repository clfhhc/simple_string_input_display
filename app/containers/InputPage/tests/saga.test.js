import { put, fork, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { INPUT_PAGE_INSERT, INPUT_PAGE_FETCH_HISTORY } from '../constants';
import { fetchedError, inputInserted, updateHistory } from '../actions';
import sagas, {
  fetchIsLoading,
  requestHistoryFromServer,
  insertStringToServer,
} from '../sagas';
import { cancel } from '../../../../node_modules/redux-saga/lib/internal/io';

const stringToInput = 'test';

/* eslint-disable redux-saga/yield-effects */
describe('insertStringToServer Saga', () => {
  let insertStringToServerGenerator;
  let forkDescriptor;

  beforeEach(() => {
    insertStringToServerGenerator = insertStringToServer();

    const selectDescriptor = insertStringToServerGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    forkDescriptor = insertStringToServerGenerator.next(stringToInput).value;
    expect(forkDescriptor).toMatchSnapshot();
  });

  it('should fork fetchIsLoading generator when loading', () => {
    const mockFork = fork(fetchIsLoading);
    expect(forkDescriptor).toEqual(mockFork);
  });

  it('should dispatch the inputInserted action if it posts the data successfully', () => {
    const mockFork = fork(fetchIsLoading);
    const mockForkTask = createMockTask(mockFork);
    insertStringToServerGenerator.next(mockForkTask);
    const cancelDescriptor = insertStringToServerGenerator.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockForkTask));
    const putDescriptor = insertStringToServerGenerator.next().value;
    expect(putDescriptor).toEqual(put(inputInserted()));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const mockFork = fork(fetchIsLoading);
    const mockForkTask = createMockTask(mockFork);
    insertStringToServerGenerator.next(mockForkTask);
    const response = new Error('Some error');
    const cancelDescriptor = insertStringToServerGenerator.throw(response)
      .value;
    expect(cancelDescriptor).toEqual(put(fetchedError(response)));
  });
});

describe('insertInput Saga', () => {
  const insertInputSaga = sagas.insertInput();

  it('should start task to watch for INPUT_PAGE_INSERT action', () => {
    const takeLatestDescriptor = insertInputSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(INPUT_PAGE_INSERT, insertStringToServer),
    );
  });
});

describe('requestHistoryFromServer Saga', () => {
  let requestHistoryFromServerSaga;
  let forkDescriptor;

  beforeEach(() => {
    requestHistoryFromServerSaga = requestHistoryFromServer();

    forkDescriptor = requestHistoryFromServerSaga.next().value;
    expect(forkDescriptor).toMatchSnapshot();
  });

  it('should fork fetchIsLoading generator when loading', () => {
    const mockFork = fork(fetchIsLoading);
    expect(forkDescriptor).toEqual(mockFork);
  });

  it('should dispatch the inputInserted action if it fetches the data successfully', () => {
    const mockFork = fork(fetchIsLoading);
    const mockForkTask = createMockTask(mockFork);
    requestHistoryFromServerSaga.next(mockForkTask);
    const response = ['hi', 'hello'];
    const cancelDescriptor = requestHistoryFromServerSaga.next({
      inputHistory: response,
    }).value;
    expect(cancelDescriptor).toEqual(cancel(mockForkTask));
    const putDescriptor = requestHistoryFromServerSaga.next().value;
    expect(putDescriptor).toEqual(put(updateHistory(response)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const mockFork = fork(fetchIsLoading);
    const mockForkTask = createMockTask(mockFork);
    requestHistoryFromServerSaga.next(mockForkTask);
    const response = new Error('Some error');
    const cancelDescriptor = requestHistoryFromServerSaga.throw(response).value;
    expect(cancelDescriptor).toEqual(put(fetchedError(response)));
  });
});

describe('fetchHistory Saga', () => {
  const fetchHistorySaga = sagas.fetchHistory();

  it('should start task to watch for INPUT_PAGE_FETCH_HISTORY action', () => {
    const takeLatestDescriptor = fetchHistorySaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(INPUT_PAGE_FETCH_HISTORY, requestHistoryFromServer),
    );
  });
});
