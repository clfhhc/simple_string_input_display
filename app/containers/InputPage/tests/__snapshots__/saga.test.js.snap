// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertStringToServer Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`insertStringToServer Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`insertStringToServer Saga should dispatch the inputInserted action if it posts the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`insertStringToServer Saga should dispatch the inputInserted action if it posts the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`insertStringToServer Saga should fork fetchIsLoading generator when loading 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`insertStringToServer Saga should fork fetchIsLoading generator when loading 2`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`requestHistoryFromServer Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`requestHistoryFromServer Saga should dispatch the inputInserted action if it fetches the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`requestHistoryFromServer Saga should fork fetchIsLoading generator when loading 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
